src/state.rs

```diff
@@ -38,25 +38,65 @@ impl OPRFServer {
 }
 
 /// Advance to the next epoch on a timer
-/// This can be invoked as a background task to handle
-/// epoch advance and key rotation according to the
-/// given Config.
+/// This can be invoked as a background task to handle epoch
+/// advance and key rotation according to the given Config.
 #[instrument(skip_all)]
 pub async fn epoch_loop(state: OPRFState, config: &Config) {
     let interval =
         std::time::Duration::from_secs(config.epoch_seconds.into());
     info!("rotating epoch every {} seconds", interval.as_secs());
 
+    let starttime = time::OffsetDateTime::now_utc();
+    // Parse the epoch basetime if given.
+    let basetime: Option<time::OffsetDateTime> = config.into();
+    // If no epoch basetime was specified, use the startup time.
+    let basetime = basetime.unwrap_or(starttime);
```

Comment:
Why unix epoch in particular?

Reply:
just out of convenience. the time crate already defines a `UNIX_EPOCH` constant

Reply:
My objection is that all of these configuration switches are about tuning the config for testing and experimentation. Having a fixed base time by default means running with `--first-epoch` is surprising because it (usually) won't start with the given value. Doing the naive thing and requesting and evaluation with that epoch will fail!

Anything that's surprising is extra cognitive load for someone trying out the software for the first time. Maybe it actually makes sense for their application to have a fixed rotation schedule, like it does for ours, but they have to go track down why it works that way, update their own work to accommodate, etc. So it's not a good default.

On the other hand, if we're hard-coding a fixed base time for our application, we don't even need the switch; we could just *always* use `UNIX_EPOCH` or whatever, and have less code to maintain. The less-surprise issue justifies having the switch, but only if it's optional.

Reply:
good, simple documentation can address the above concerns. if we indicate that the first & last epoch switches are used for defining the full range of possible epochs, then i think that could mitigate any potential confusion.

> Doing the naive thing and requesting and evaluation with that epoch will fail!

if they are querying the info endpoint, or they are paying attention to the logs, they should know the current epoch

> On the other hand, if we're hard-coding a fixed base time for our application, we don't even need the switch; we could just always use UNIX_EPOCH or whatever, and have less code to maintain.

your previous points indicate that you're concerned about third-party users. i think keeping the switch will be good for such users who want to adhere to a different epoch schedule, as well as ad-hoc testing. i'm not sure if code maintenance is much of a concern, since clap would be doing most of the heavy lifting anyway

Reply:
also, it seems strange that we provide a hardcoded switch in the start script: https://github.com/brave/star-randsrv/blob/02837597eeed8b0b3625b7a9b3dbabcabfc3e5e3/start.sh#L16

but if a dev (a first-party dev, more importantly) runs the server locally using `cargo run`, they will not get the same behavior by default. that could result in more time troubleshooting. we could be shooting ourselves in the foot by not providing a sensible default.

Reply:
> good, simple documentation can address the above concerns. [...] if they are querying the info endpoint, or they are paying attention to the logs, they should know the current epoch

Those are precisely the things I was referring to as extra cognitive load.

> we provide a hardcoded switch in the start script

Agree with you there. That's our need to fix all the build inputs for reproducibility running up against not surprising defaults.

@NullHypothesis, do you have an opinion?

Reply:
>Agree with you there. That's our need to fix all the build inputs for reproducibility running up against not surprising defaults. Maybe we can do something about that later with a default value and env override like we were talking about for the fqdn.

We could separate star-randsrv's code from the nitriding-relevant build scripts (in particular start.sh and Dockerfile). If we're concerned about usability for third parties, we shouldn't impose a specific way to run star-randsrv.

Reply:
> Maybe we can do something about that later with a default value and env override like we were talking about for the fqdn.

were you thinking of adding the default value in start.sh, the Dockerfile, or via `--env` switch in the Docker build? if creating a default is the solution to the concern, perhaps it would be best to just keep it in the server itself. it's more visible for others, especially those who are not running the server in a Docker image or via start.sh, i.e. developers that are compiling and running via `cargo run`. 

to override the variable, we can simply pass the arguments of the `start.sh` script directly to the server, and specify an override by passing in a `--env` switch to Docker during build time.

Reply:
> If we're concerned about usability for third parties, we shouldn't impose a specific way to run star-randsrv.

if we adapt start.sh or the Dockerfile to pass in additional arguments to the server if a relevant build time environment variable (i.e. something like `EPOCH_BASE_TIME`) exists, then we should be able to provide some extensibility for third-party users.

Reply:
thoughts @rillian ?

Reply:
We could move those files to a `enclave` subdir if the top-level Dockerfile is confusing.

Reply:
we will have to expose some parameters in `start.sh` / `Dockerfile` for our own purposes soon, i.e. for different reporting cadences/epoch lengths. which probably means that we will need to keep our build config/parameter values in the github build config (which lives in this repo, luckily).

my main question is about where defaults should live. my rationale for having them live within clap was stated above:

> it's more visible for others, especially those who are not running the server in a Docker image or via start.sh, i.e. developers that are compiling and running locally via cargo run.

Reply:
not really a showstopper, just something to think about